### DISCUSSION TITLE
Non-openGL related class as a plugin of the viewer?

### DISCUSSION THREAD

**Discussion by xarthurx** [libigl/libigl#1758] [Q&A] [answered]
This related to an earlier question I asked: https://github.com/libigl/libigl/issues/1161

I understand that the plugin system in libigl can be used to extend the functionality of the default ones.
However, as my project goes complex, using only one `main.cpp` file is a bit complicated and I would like to separate code into different `.h` and `.cpp` files for better organization of the code.

For that, I defined different classes in different header files, and included them in the different `.cpp` file.
The question occurs when an object needs to be used in multiple class (I currently use a `shared_ptr` for that), and I'm not sure what should be holding that object.

In my previous project where libigl is not used, I put all related staff in a `viewer` class, so that any functions defined in the viewer or called in the `main.cpp` can use it. 
For `libigl`, it seems the `Viewer` class itself should not be inherited, and I guess that's why the `plugin` system is developed?

My question is, am I supposed to derive my personal class object from the `Viewer` class if I need it to be used in the `main.cpp` or from inside the `viewer`? No matter if it is openGL related or not?
From what I understand, the `plugin` base class is just an empty class that hold several virtual functions, which are pre-written for ease of use?

If not, what is the correct way or scale the project up based on `libigl`?

**Comment by jdumas** [comment, marked as answer]
A general strategy that I use is the following:
- I try to split my project in a "lib" part and an "app" part. The lib contains free functions in a libigl-style that operate on the data I need. I pass explicitly all the arguments that I need to theses functions.
- The "app" part contains the viewer, organized in one main plugin, and two "state" classes.
- The two classes `State` and `UIState` that are just containers with data.
- `State` contains e.g. my scene, and any other object that I would need to run my code in "headless" mode (without the viewer)
- `UIState` contains any data that I need for the viewer (e.g. a set of ImGui options, selected colorbar, etc.)
- The main class `MyApp` derives from `igl::opengl::glfw::imgui::ImGuiMenu`. This class contains all the methods I need to run my app. The implementation can be split across several .cpp, but is generally quite straightforward as it just aggregates function calls from my main "lib".

You can find example of a similar strategy on the [PolyFEM](https://github.com/cellogram/cellogram/tree/master/app/viewer) and [Cellogram](https://github.com/cellogram/cellogram/tree/master/app/viewer) projects, although those use a single `UIState` class which is both the data container and the viewer plugin.

**Reply by xarthurx** [reply]
This is indeed helpful! Thank you!
